## MICE Algorithm

With the help of following steps of MICE algorithm, we can find out the relation of a varibale with other variables and use this relationship to impute the missing values.

1. Remove all the missing values of the column let say col_1 for which imputation is required.
2. Predict the misssing values of col_1 using other columns
3. In the similar way remove all the missing values of column 2.
4. Predict the missing values of column 2 using other columns.
5. Do iterations.

Let's perform all these steps with the help of an example.

In this example we are going to use 50_Startups dataset. You can access the dataset [here](https://drive.google.com/file/d/1PZq_5_sCFMbW_W8AZXUuj5xYhVETLBMg/view?usp=share_link
)

In [ ]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LinearRegression

We are going to fetch 5 samples from the dataset and work on those.

In [ ]:
df = np.round(pd.read_csv('50_Startups.csv')[['R&D Spend','Administration','Marketing Spend','Profit']]/10000)
np.random.seed(9)
df = df.sample(5)
df

R&D Spend  Administration  Marketing Spend  Profit
21        8.0            15.0             30.0    11.0
37        4.0             5.0             20.0     9.0
2        15.0            10.0             41.0    19.0
14       12.0            16.0             26.0    13.0
44        2.0            15.0              3.0     7.0

Now drop the Profit column because it is not required for the imputation.

In [ ]:
df = df.iloc[:,0:-1]
df

R&D Spend  Administration  Marketing Spend
21        8.0            15.0             30.0
37        4.0             5.0             20.0
2        15.0            10.0             41.0
14       12.0            16.0             26.0
44        2.0            15.0              3.0

As the dataset does not contain any null values, we are going to randomly introduce the null values in some places.

In [ ]:
df.iloc[1,0] = np.NaN
df.iloc[3,1] = np.NaN
df.iloc[-1,-1] = np.NaN

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)
/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:723: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value, self.name)


In [ ]:
df.head()

R&D Spend  Administration  Marketing Spend
21        8.0            15.0             30.0
37        NaN             5.0             20.0
2        15.0            10.0             41.0
14       12.0             NaN             26.0
44        2.0            15.0              NaN

In [ ]:
df0 = pd.DataFrame()

df0['R&D Spend'] = df['R&D Spend'].fillna(df['R&D Spend'].mean())
df0['Administration'] = df['Administration'].fillna(df['Administration'].mean())
df0['Marketing Spend'] = df['Marketing Spend'].fillna(df['Marketing Spend'].mean())

In [ ]:
df0

R&D Spend  Administration  Marketing Spend
21       8.00           15.00            30.00
37       9.25            5.00            20.00
2       15.00           10.00            41.00
14      12.00           11.25            26.00
44       2.00           15.00            29.25

You can see that if we fill the null values using the mean imputation, they vary so much from the original values.

In the **Administration** colum, it can be seen that the missing column is replaced with 11.25 but the original value is 16. 

Now let's see how MICE method can help in getting the correct value after imputation.

We are creating a new dataframe which is the copy of original sample. Then introduce a NaN value in the **R&D Spend** column.

In [ ]:
df1 = df0.copy()

df1.iloc[1,0] = np.NaN

df1

R&D Spend  Administration  Marketing Spend
21        8.0           15.00            30.00
37        NaN            5.00            20.00
2        15.0           10.00            41.00
14       12.0           11.25            26.00
44        2.0           15.00            29.25

We are going to create a X variable which holds all the independent variables for which **R&D Spend** is not null. This is because we want to train a regressor on these values.

In [ ]:
X = df1.iloc[[0,2,3,4],1:3]
X

Administration  Marketing Spend
21           15.00            30.00
2            10.00            41.00
14           11.25            26.00
44           15.00            29.25

And y variable which holds all the non null values of **R&D Spend** column.

In [ ]:
y = df1.iloc[[0,2,3,4],0]
y

21     8.0
2     15.0
14    12.0
44     2.0
Name: R&D Spend, dtype: float64

Now train a linear regressor on the values of X and y.

In [ ]:
lr = LinearRegression()
lr.fit(X,y)
lr.predict(df1.iloc[1,1:].values.reshape(1,2))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([23.14158651])

In [ ]:
df1.iloc[1,0] = 23.14

In [ ]:
df1

R&D Spend  Administration  Marketing Spend
21       8.00           15.00            30.00
37      23.14            5.00            20.00
2       15.00           10.00            41.00
14      12.00           11.25            26.00
44       2.00           15.00            29.25

We are going to perform the same steps for **Administration** column now.

In [ ]:
df1.iloc[3,1] = np.NaN

df1

R&D Spend  Administration  Marketing Spend
21       8.00            15.0            30.00
37      23.14             5.0            20.00
2       15.00            10.0            41.00
14      12.00             NaN            26.00
44       2.00            15.0            29.25

In [ ]:
X = df1.iloc[[0,1,2,4],[0,2]]
X

R&D Spend  Marketing Spend
21       8.00            30.00
37      23.14            20.00
2       15.00            41.00
44       2.00            29.25

In [ ]:
y = df1.iloc[[0,1,2,4],1]
y

21    15.0
37     5.0
2     10.0
44    15.0
Name: Administration, dtype: float64

In [ ]:
lr = LinearRegression()
lr.fit(X,y)
lr.predict(df1.iloc[3,[0,2]].values.reshape(1,2))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([11.06331285])

In [ ]:
df1.iloc[3,1] = 11.06

In [ ]:
df1

R&D Spend  Administration  Marketing Spend
21       8.00           15.00            30.00
37      23.14            5.00            20.00
2       15.00           10.00            41.00
14      12.00           11.06            26.00
44       2.00           15.00            29.25

In [ ]:
df1.iloc[4,-1] = np.NaN

df1

R&D Spend  Administration  Marketing Spend
21       8.00           15.00             30.0
37      23.14            5.00             20.0
2       15.00           10.00             41.0
14      12.00           11.06             26.0
44       2.00           15.00              NaN

In [ ]:
X = df1.iloc[0:4,0:2]
X

R&D Spend  Administration
21       8.00           15.00
37      23.14            5.00
2       15.00           10.00
14      12.00           11.06

In [ ]:
y = df1.iloc[0:4,-1]
y

21    30.0
37    20.0
2     41.0
14    26.0
Name: Marketing Spend, dtype: float64

In [ ]:
lr = LinearRegression()
lr.fit(X,y)
lr.predict(df1.iloc[4,0:2].values.reshape(1,2))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([31.56351448])

In [ ]:
df1.iloc[4,-1] = 31.56

In [ ]:
df1

R&D Spend  Administration  Marketing Spend
21       8.00           15.00            30.00
37      23.14            5.00            20.00
2       15.00           10.00            41.00
14      12.00           11.06            26.00
44       2.00           15.00            31.56

In [ ]:
df1 - df0

R&D Spend  Administration  Marketing Spend
21       0.00            0.00             0.00
37      13.89            0.00             0.00
2        0.00            0.00             0.00
14       0.00           -0.19             0.00
44       0.00            0.00             2.31

In [ ]:
df2 = df1.copy()

df2.iloc[1,0] = np.NaN

df2

R&D Spend  Administration  Marketing Spend
21        8.0           15.00            30.00
37        NaN            5.00            20.00
2        15.0           10.00            41.00
14       12.0           11.06            26.00
44        2.0           15.00            31.56

In [ ]:
X = df2.iloc[[0,2,3,4],1:3]
y = df2.iloc[[0,2,3,4],0]

lr = LinearRegression()
lr.fit(X,y)
lr.predict(df2.iloc[1,1:].values.reshape(1,2))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([23.78627207])

In [ ]:
df2.iloc[1,0] = 23.78

In [ ]:
df2.iloc[3,1] = np.NaN
X = df2.iloc[[0,1,2,4],[0,2]]
y = df2.iloc[[0,1,2,4],1]

lr = LinearRegression()
lr.fit(X,y)
lr.predict(df2.iloc[3,[0,2]].values.reshape(1,2))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([11.22020174])

In [ ]:
df2.iloc[3,1] = 11.22

In [ ]:
df2.iloc[4,-1] = np.NaN

X = df2.iloc[0:4,0:2]
y = df2.iloc[0:4,-1]

lr = LinearRegression()
lr.fit(X,y)
lr.predict(df2.iloc[4,0:2].values.reshape(1,2))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([38.87979054])

In [ ]:
df2.iloc[4,-1] = 38.87

In [ ]:
df2

R&D Spend  Administration  Marketing Spend
21       8.00           15.00            30.00
37      23.78            5.00            20.00
2       15.00           10.00            41.00
14      12.00           11.22            26.00
44       2.00           15.00            38.87

In [ ]:
df2 - df1

R&D Spend  Administration  Marketing Spend
21       0.00            0.00             0.00
37       0.64            0.00             0.00
2        0.00            0.00             0.00
14       0.00            0.16             0.00
44       0.00            0.00             7.31